### PR TITLE
test: reduce test with inmemory watermark store

### PR DESCRIPTION
**Pull request by yhl25**
Signed-off-by: Yashash H L <yashashhl25@gmail.com>

reduce tests integrated with in-memory watermark store

### CODE REVIEW

**Review by whynowy — APPROVED** [review]
LGTM

File: pkg/reduce/reduce_test.go [hunk 1]
```diff
@@ -143,26 +146,107 @@ func TestDataForward_Start(t *testing.T) {
...
+	hb, hbWatcherCh, _ := inmem.NewKVInMemKVStore(ctx, pipelineName, hbBucketName)
+	ot, otWatcherCh, _ := inmem.NewKVInMemKVStore(ctx, pipelineName, otBucketName)
+
+	publishers := make(map[string]publish.Publisher)
+
+	p := publish.NewPublish(ctx, publishEntity, wmstore.BuildWatermarkStore(hb, ot), publish.WithAutoRefreshHeartbeatDisabled(), publish.WithPodHeartbeatRate(1))
+	sourcePublisher := publish.NewPublish(ctx, sourcePublishEntity, wmstore.BuildWatermarkStore(hb, ot), publish.WithAutoRefreshHeartbeatDisabled(), publish.WithPodHeartbeatRate(1))
+
+	go func() {
+		for {
```
Comment (vigith) [inline pkg/reduce/reduce_test.go:232]: this go routine will be leaked. Can we use context or something to exit out?

Comment (vigith) [reply pkg/reduce/reduce_test.go:232]: use context with cancelFn and call cancel via defer

File: pkg/reduce/reduce_test.go [hunk 2]
```diff
@@ -143,26 +146,107 @@ func TestDataForward_Start(t *testing.T) {
...
+	sourcePublishEntity := processor.NewProcessorEntity("sourceProcessor")
+	hb, hbWatcherCh, _ := inmem.NewKVInMemKVStore(ctx, pipelineName, hbBucketName)
+	ot, otWatcherCh, _ := inmem.NewKVInMemKVStore(ctx, pipelineName, otBucketName)
+
+	publishers := make(map[string]publish.Publisher)
+
+	p := publish.NewPublish(ctx, publishEntity, wmstore.BuildWatermarkStore(hb, ot), publish.WithAutoRefreshHeartbeatDisabled(), publish.WithPodHeartbeatRate(1))
+	sourcePublisher := publish.NewPublish(ctx, sourcePublishEntity, wmstore.BuildWatermarkStore(hb, ot), publish.WithAutoRefreshHeartbeatDisabled(), publish.WithPodHeartbeatRate(1))
+
+	go func() {
```
Comment (vigith) [inline pkg/reduce/reduce_test.go:231]: why do we need this? can't we put watermark while writing messages?

File: pkg/reduce/reduce_test.go (hunk 1 above)
Comment (yhl25) [reply pkg/reduce/reduce_test.go:362]: done

File: pkg/reduce/reduce_test.go (hunk 2 above)
Comment (yhl25) [reply pkg/reduce/reduce_test.go:361]: this is for heartbeat

File: pkg/reduce/reduce_test.go [hunk 3]
```diff
@@ -73,14 +77,72 @@ func (f forwardReduceTest) Reduce(ctx context.Context, messageStream <-chan *isb
...
+		_ = json.Unmarshal(msg.Payload, &payload)
+		sum += payload.Value
+	}
+
+	payload := PayloadForTest{Key: "sum", Value: sum}
+	b, _ := json.Marshal(payload)
+	ret := &isb.Message{
+		Header: isb.Header{
+			PaneInfo: isb.PaneInfo{
+				EventTime: time.Now(),
```
Comment (vigith) [inline pkg/reduce/reduce_test.go:100]: we need to pass in window endtime

**Review by vigith — APPROVED** [review]

**Pull request merged**